### DISCUSSION TITLE
is it possible to fill pinia in server-side on nuxt 3?

### DISCUSSION THREAD

**Discussion by ghost** [vuejs/pinia#947] [Help and Questions] [answered]
`store/todos.ts`
```ts
import { defineStore } from 'pinia'
import axios from '@/utils/axios'

export const useTodos = defineStore('todos', {
  state: () => ({
    todo: {}
  }),

  actions: {
    async fetch() {
      const { data } = await axios.get('/todos/1')

      this.todo = data
    }
  }
})
```

`plugins/init.server.ts`
```ts
import { useTodos } from '@/store/todos'

const todos = useTodos()

todos.fetch()
```

![image info](https://i.ibb.co/GCP0DmB/Screen-Shot-2022-01-10-at-19-22-03.png)

I am trying to replicate the `nuxtServerInit() function of Nuxt2`, to start populating data in components once all the data is loaded from the API.

**Comment by ghost** [comment, marked as answer]
my bad

```ts
import { defineNuxtPlugin } from '#app'
import { useTodos } from '@/store/todos'

export default defineNuxtPlugin(async (nuxtApp) => {
  const todos = useTodos(nuxtApp.$pinia)

  await todos.fetch()
})
```

**Reply by minhyentcu** [reply]
Can you share the source code? i am having the same problem but haven't solved it yet.

**Reply by Dorianslavov** [reply]
doesn't this get triggered on every route change, even in a single session including `<NuxtLink>` navigations which do not reload the page ?

**Reply by justin-schroeder** [reply]
no, that would be route middleware. this is just a plugin

**Reply by pkanane** [reply]
@Dorianslavov it does get triggered on route changes in the browser. I don't know about `<NuxtLink>` In implementing this approach, I've considered the impact. Do you have alternative suggestions?

**Reply by Dorianslavov** [reply]
@pkanane https://github.com/vuejs/pinia/discussions/947#discussioncomment-3279213

**Reply by TutorFx** [reply]
Thank you for the share! :)

**Reply by alfianriv** [reply]
@yungsilva  ```nuxtApp.$pinia type is unknown``` and parameter ```useGlobalStore``` need ```Pinia``` type. Any other solutions ?
![image](https://user-images.githubusercontent.com/16283592/233553083-72fc00bb-efd0-4853-9ba8-03e8d244d93b.png)

**Reply by abdinurkz** [reply]
same issue

**Reply by doantu05150** [reply]
I hate to do this but set `nuxtApp: any` to ignore this issue.

**Reply by alfianriv** [reply]
@AbdinurK you can try `nuxtApp.$nuxt.$pinia`

**Reply by bastianjoel** [reply]
As of [mentioned here](https://nuxt.com/docs/guide/directory-structure/plugins#typing-plugins) you should use `useNuxtApp().$pinia`.

**Reply by YadBro** [reply]
Try this:
export default defineNuxtPlugin(({ vueApp }) => {
  vueApp.$nuxt.$pinia.use(MyPiniaPlugin)
});

"nuxt": "^3.6.5"

**Comment by minhyentcu** [comment]
Can you share the source code? i am having the same problem but haven't solved it yet.

**Comment by Dorianslavov** [comment]
✅ This is the way I've made it work

Let's say in somewhere in your store you have a method `initProfile` which you want to call server-side

Call that method in `layouts/default.vue` or in all of your `/layouts/*.vue` using `useAsyncData`
example - `await useAsyncData('init', initProfile)`
```ts
// layouts/default.vue
<script setup lang="ts">
import { useUserStore } from '~/store/user'

const userStore = useUserStore()
const { isLogged, initProfile } = userStore

// Init profile on server
if (isLogged) await useAsyncData('profile', initProfile)
</script>
```

That way your high-level layout would always populate the store on the server-side

❗️The issue with this method would be that pages with `layout: false` can't call this method. For this issue, you can create your own `no-layout` ( or call it whatever you want ) layout which still calls the init method in it's `<script setup>` but includes only the `<NuxtPage />` in the `<template>` like that and use the `no-layout` layout instead of `layout: false`

```html
<template>
    <main>
        <NuxtPage />
    </main>
</template>
```

That way you would ensure that all of the pages will call `initProfile`

**Reply by briansackett** [reply]
This worked for me, so thank you! Just wanted to put this here for reference.. 
When using destructuring refs from a store you should be using storeToRefs() so they are reactive.. not required for functions.

`
import { storeToRefs } from 'pinia';
const userStore = useUserStore();
const { isLogged } = storeToRefs(userStore);
const {  initProfile } = userStore();
`

**Comment by melokki** [comment]
Hello, I tried the above implementation, but the function in store is executed every time I click on a `<NuxtLink>`
When I check the length of the store array property, so I can stop its execution, the length is always 0 despite the fact that it was previously filled with data that I can see in `devtools`. 
I don't understand how to make this work

Does anyone have some ideas?

I use `rc.11` version of nuxt

**Comment by MuhammadM1998** [comment]
How can I do it with a nitro plugin (putting my code in `/server/plugins` ?

**Reply by SuperIbm** [reply]
Did you find any solution to do this? I want to elaborate why I personally need this. I want to use [Server Storage](https://nuxt.com/docs/guide/directory-structure/server#server-storage) in order to cache my data via redis. But I do not know how to combine pinia + nuxt + axios + nitro plugin all together. The problem is that you are not able to use useStorage out of defineEventHandler it means that you can't just use useStorage in the pinia side. Also I tried to include Redis directly https://github.com/redis/node-redis and https://github.com/luin/ioredis in order to cache data with my own solution, but it didn't work with Nuxt 3 also.

**Reply by MuhammadM1998** [reply]
I had no luck sorry

**Comment by LeonardoRick** [comment]
Hey, is this still possbile? i'm not being able to call composables inside plugins at all
```

[nuxt] [request error] [unhandled] [500] Must be called at the top of a setup function                                                                                                               4:26:29 PM
  at Module.createCompileError (./node_modules/@intlify/message-compiler/dist/message-compiler.mjs:54:19)  
  at createI18nError (./node_modules/vue-i18n/dist/vue-i18n.mjs:97:34)  
  at Module.useI18n (./node_modules/vue-i18n/dist/vue-i18n.mjs:2248:15)  
  at ./store/index.ts:12:33  
  at EffectScope.run (./node_modules/@vue/reactivity/dist/reactivity.cjs.js:42:16)  
  at ./node_modules/pinia/dist/pinia.mjs:1438:43  
  at runWithContext (./node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:3855:18)  
  at ./node_modules/pinia/dist/pinia.mjs:1438:16  
  at EffectScope.run (./node_modules/@vue/reactivity/dist/reactivity.cjs.js:42:16)  
  at createSetupStore (./node_modules/pinia/dist
```

**Comment by johannschopplich** [comment]
You can also use [`nuxt-prepare`](https://nuxt-prepare.byjohann.dev/) if it is enough for you to fetch data at build time (not runtime).

I have written an article on how to [pass state from a prepare script to Nuxt](https://nuxt-prepare.byjohann.dev/guide/example-pass-state-to-nuxt.html).

In a nutshell, you can import your prefetched data anywhere from the `#nuxt-prepare` import:

```ts
// `stores/todo.ts`
import { defineStore } from 'pinia'
import { todos } from '#nuxt-prepare'

export const useTodos = defineStore('todos', {
  state: () => ({
    todos: todos || [],
  })
})
```

For the prepare script, see the docs. It basically boils down to:

```ts
// `store.prepare.ts`
import { $fetch } from 'ofetch'
import { defineNuxtPrepareHandler } from '../src/config'

export default defineNuxtPrepareHandler(async () => {
  const todos = await $fetch('todos/1', {

  return {
    // Pass your todos data to Nuxt and import it
    // anywhere from `#nuxt-prepare`
    state: {
      todos,
    },
  }
})

```
